components/src/jsMain/kotlin/dev/fritz2/components/validation/validation.kt

```diff
@@ -0,0 +1,38 @@
+package dev.fritz2.components.validation
+
+import dev.fritz2.binding.RootStore
+import dev.fritz2.binding.Store
+import dev.fritz2.binding.SubStore
+import dev.fritz2.validation.Validator
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+
+interface Validatable<D, T> : Store<D> {
+    val validator: Validator<D, ComponentValidationMessage, T>
+    fun validate(metadata: T) {
+        syncBy(handle<D> { old, new ->
+            if (validator.isValid(new, metadata)) new else old
+        })
+    }
+}
+
+class RootStoreIsNotValidatable(id: String) :
+    Exception("RootStore of data with id=$id must implement Validatable interface")
+
+fun <D> Store<D>.validationMessage(): Flow<String> = when (this) {
+    is RootStore<*> -> {
+        if (this is Validatable<*, *>) {
+            this.validator.find { it.id == this@validationMessage.id }.map { it?.message ?: "" }
```

Comment:
What about returning all messages for a given id here and leave it to the user to take just the first if needed? Or offer both methods? (validationMessage and validationMessages)

Reply:
Yes, good point. I change the returning type of `validationMessage` function to `Flow<ComponentValidationMessage?>` which more general and can be mapped afterwards. Additionally I have added a second extension method `validationMessages` which return a `Flow<List<ComponentValidationMessage>>` in case you have more then one message for field in your data model.